File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    String[] map = {
            " ",
            " ",
            "abc",
            "def",
            "ghi",
            "jkl",
            "mno",
            "pqrs",
            "tuv",
            "wxyz"
        };

    List<String> ans;

    private void backtrack(String digits , int idx , StringBuilder sb){
        if(idx == digits.length()){
            ans.add(sb.toString());
            return;
        }

        String curr = map[digits.charAt(idx) - '0'];

        for(char ch : curr.toCharArray()){
            sb.append(ch);
            backtrack(digits , idx + 1 , sb);
            sb.deleteCharAt(sb.length() - 1);
        }
    }

    public List<String> letterCombinations(String digits) {
        ans = new ArrayList<>();
        StringBuilder sb = new StringBuilder();

        backtrack(digits , 0 , sb);

        return ans;
    }
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:

    def __init__(self):
        self.ans = []
        self.mp = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz"
        }

    def letterCombinations(self, digits: str) -> List[str]:

        if not digits:
            return []

        def backtrack(idx , path):
            if len(path) == len(digits):
                self.ans.append("".join(path))
                return

            for ch in self.mp[digits[idx]]:
                path.append(ch)
                backtrack(idx + 1 , path)
                path.pop()

        backtrack(0 , [])
        return self.ans
